File: LC009PalindromeNumber.java
package leetcode;

/**
 * https://leetcode.com/problems/palindrome-number/
 */
public class LC009PalindromeNumber {
    public boolean isPalindrome(int x) {
        if (x < 0) {
            return false;
        }
        if (x == 0) {
            return true;
        }
        int power = (int) (Math.floor(Math.log10(x)));
        int m = (int) Math.pow(10, power);
        for (int i = 0; i < Math.ceil((power + 1) / 2.0); i++) {
            x -= (x % 10) * m;
            if (x < 0) {
                return false;
            }
            x /= 10;
            m /= 100;
        }
        return x == 0;
    }

    public static void main(String[] arguments) {
        LC009PalindromeNumber solution = new LC009PalindromeNumber();
        System.out.println(solution.isPalindrome(213));
        System.out.println(solution.isPalindrome(456));
        System.out.println(solution.isPalindrome(454));
        System.out.println(solution.isPalindrome(99));
        System.out.println(solution.isPalindrome(1));
        System.out.println(solution.isPalindrome(10));
    }
}

File: algorithm-design/leetcode/lc009_palindrome_number/lc009_palindrome_number.py
#!/usr/bin/env python3
# https://leetcode.com/problems/palindrome-number/

import unittest


class Solution:
    def isPalindrome(self, x: int) -> bool:
        if x < 0:
            return False
        xa = x
        xb = 0
        while x > 0:
            xb = xb * 10 + x % 10
            x //= 10
        return xa == xb


class TestCode(unittest.TestCase):
    def test_213(self):
        self.assertEqual(Solution().isPalindrome(213), False)

    def test_456(self):
        self.assertEqual(Solution().isPalindrome(456), False)

    def test_454(self):
        self.assertEqual(Solution().isPalindrome(454), True)

    def test_99(self):
        self.assertEqual(Solution().isPalindrome(99), True)

    def test_1(self):
        self.assertEqual(Solution().isPalindrome(1), True)

    def test_10(self):
        self.assertEqual(Solution().isPalindrome(10), False)

    def test_minus1(self):
        self.assertEqual(Solution().isPalindrome(-1), False)

    def test_0(self):
        self.assertEqual(Solution().isPalindrome(0), True)
